zkevm-circuits/src/evm_circuit/execution/mul.rs

```diff
@@ -29,30 +33,66 @@ impl<F: Field> ExecutionGadget<F> for MulGadget<F> {
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
 
-        let a = cb.query_word();
-        let b = cb.query_word();
+        let is_mul = (OpcodeId::DIV.expr() - opcode.expr())
+            * (OpcodeId::MOD.expr() - opcode.expr())
+            * F::from(8).invert().unwrap();
+        let is_div = (opcode.expr() - OpcodeId::MUL.expr())
+            * (OpcodeId::MOD.expr() - opcode.expr())
+            * F::from(4).invert().unwrap();
+        let is_mod = (opcode.expr() - OpcodeId::MUL.expr())
+            * (opcode.expr() - OpcodeId::DIV.expr())
+            * F::from(8).invert().unwrap();
+        let mul_add_words = MulAddWordsGadget::construct(cb);
 
-        // Pop a and b from the stack, push product on the stack
-        cb.stack_pop(a.expr());
-        cb.stack_pop(b.expr());
-        let mul_words = MulWordsGadget::construct(cb, a, b);
-        cb.stack_push(mul_words.product().expr());
+        // Pop a and b from the stack, push result on the stack
+        // The first pop is multiplier for MUL and dividend for DIV/MOD
+        // The seoncd pop is multiplicand for MUL and divosr for DIV/MOD
+        // The push is product for MUL, quotient for DIV, and residue for MOD
+        cb.stack_pop(select::expr(
+            is_mul.clone(),
+            mul_add_words.a.expr(),
+            mul_add_words.d.expr(),
+        ));
+        cb.stack_pop(mul_add_words.b.expr());
+        cb.stack_push(
+            is_mul.clone() * mul_add_words.d.expr()
+                + is_div * mul_add_words.a.expr()
+                + is_mod * mul_add_words.c.expr(),
+        );
+
+        // Constraint for MUL case
+        cb.add_constraint(
+            "c == 0 for opcode MUL",
+            is_mul.clone() * sum::expr(&mul_add_words.c.cells),
```

Comment:
I think we can still use the expression to derive `is_mul` unless the degree of constraints is too high. What do you think?

Reply:
But even if you derive it, wouldn't it be possible that it is not `0` or `1`? 
I mean, you're kinda using it as a selector, and so, it would make sense that we constrain it to be within `[0,1]`. 
Maybe this is done earlier or something similar. In which case, ignore this comment/discussion.

Otherways, I think we need to constrain it. Could you explain how deriving from the expression will guarantee that the value is within the correct range?

Reply:
`is_mul`, `is_div` and `is_mod` can only be 0 or 1. This is because we have constrained the `opcode` to be chosen from the set `(MUL, DIV, MOD)` through the [`ResponsibleOpcode` lookup](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L37-L47).

Reply:
Thanks! That clarifies the point completely!